tests/integration_tests/functional/test_cpu_features.py

```diff
@@ -314,6 +346,367 @@ def _test_cpu_rdmsr(context):
+    This is the first part of the test verifying
+    that MSRs retain their values after restoring from a snapshot.
+
+    This function makes MSR value modifications according to the
+    ../resources/tests/msr/wrmsr_list.txt file.
+
+    Before taking a snapshot, MSR values are dumped into a text file.
+    After restoring from the snapshot on another instance, the MSRs are
+    dumped again and their values are compared to previous.
+    Some MSRs are not inherently supposed to retain their values, so they
+    form an MSR exception list.
+
+    This part of the test is responsible for taking a snapshot and publishing
+    its files along with the `before` MSR dump.
+
+    @type: functional
+    """
+    shared_names = SNAPSHOT_RESTORE_SHARED_NAMES
+
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    microvm_artifacts = ArtifactSet(
+        artifacts.microvms(keyword=shared_names["microvm_keyword"])
+    )
+    kernel_artifacts = ArtifactSet(artifacts.kernels())
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword=shared_names["disk_keyword"]))
+    assert len(disk_artifacts) == 1
+
+    snapshot_artifacts_root_dir = shared_names["snapshot_artifacts_root_dir_wrmsr"]
+    shutil.rmtree(snapshot_artifacts_root_dir, ignore_errors=True)
+    os.makedirs(snapshot_artifacts_root_dir)
+
+    test_context = TestContext()
+    test_context.custom = {
+        "builder": MicrovmBuilder(bin_cloner_path),
+        "cpu_template": cpu_template,
+        "shared_names": shared_names,
+    }
+
+    test_matrix = TestMatrix(
+        context=test_context,
+        artifact_sets=[microvm_artifacts, kernel_artifacts, disk_artifacts],
+    )
+    test_matrix.run_test(_test_cpu_wrmsr_snapshot)
+
+
+def diff_msrs(before, after, column_to_drop):
+    """
+    Calculates and formats a diff between two MSR tables.
+    """
+    # Drop irrelevant column
+    before_stripped = before.drop(column_to_drop, axis=1)
+    after_stripped = after.drop(column_to_drop, axis=1)
+
+    # Check that values in remaining columns are the same
+    all_equal = (before_stripped == after_stripped).all(axis=None)
+
+    # Arrange the diff as a side by side comparison of statuses
+    not_equal = (before_stripped != after_stripped).any(axis=1)
+    before_stripped.columns = ["MSR_ADDR", "Before"]
+    after_stripped.columns = ["MSR_ADDR", "After"]
+    diff = pd.merge(
+        before_stripped[not_equal],
+        after_stripped[not_equal],
+        on=["MSR_ADDR", "MSR_ADDR"],
+    ).to_string()
+
+    # Return the diff or an empty string
+    return diff if not all_equal else ""
+
+
+def check_msr_values_are_equal(before_msrs_fname, after_msrs_fname, guest_kernel_name):
+    """
+    Checks that MSR statuses and values in the files are equal.
+    """
+    before = pd.read_csv(before_msrs_fname)
+    after = pd.read_csv(after_msrs_fname)
+
+    # pylint: disable=singleton-comparison
+    flt_before = before[before["MSR_ADDR"].isin(MSR_EXCEPTION_LIST) == False]
```

Comment:
I think in pandas it is more idiomatic to write `before[~before[].isin()]`, and should get rid of the pylint exception.

Reply:
done